[DXIL] Consume Metadata Analysis information in passes

bharadwajy opened the pull request:
- Changed `DXILTranslateMetadata::translateMetadata()` to consume DXIL Metadata Analysis information. Subsumed into `DXILTranslateMetedata.cpp` the functionality  in `DXILMetadata.*` files - that are hence deleted.
- Changed `DXILPrepare` pass to consume DXIL Metadata Analysis information.
- Renamed `ModuleMetadataInfo::ShaderStage` to `ModuleMetadataInfo::ShaderProfile` to better convey what it represents.
- Updated `unknown` target shader stage specification in triples of a couple of tests.
- Added new tests for additional verification of `DXILTranslateMetadata` pass functionality.

Commit:
[DXIL] Consume Metadata Analysis information in DXILTranslateMetadata
and DXILPrepare passes.

python3kgae commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,273 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+      // only one entry function in the module.
+      ShaderEntryMDInfo SEP(EntryProp, Ctx, MDAnalysisInfo.ShaderProfile,
+                            MDResources, ShaderFlags);
+      MDTuple *EmptyMDT = SEP.getAsMetadata();
```
Not EmptyMDT.
Maybe CSMDT?

bharadwajy replied:
> Not EmptyMDT. Maybe CSMDT?

Missed editing after copy-paste. Thanks for catching that. Fixed.

Commit:
Fix mis-named variable that was not edited appropriately after
a copy-paste.

Add necessary function attributes to tests and remove
unnecessary check for mumber of entries for non-library
shaders in createEntryMD()

bharadwajy commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,273 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+      // EntryPro.Entry is 0.
+      ShaderEntryMDInfo SEP(EntryProp, Ctx, MDAnalysisInfo.ShaderProfile,
+                            nullptr, 0);
+      MDTuple *EmptyMDT = SEP.getAsMetadata();
```
> It is not EmptyMDT anymore.

Missed editing after copy-paste. Thanks for catching that. Fixed.

>  And we might need to test for more than one entry in lib profile.

This loop-body is intended to handle all the entries of a shader compiled for lib target profile. Did you mean something else?

python3kgae replied:
I mean we need a .ll test which checks multiple entry metadata generated correctly.

bharadwajy replied:
> I mean we need a .ll test which checks multiple entry metadata generated correctly.

I already added such a [test](https://github.com/llvm/llvm-project/blob/bf684034844c660b778f0eba103582f582b710c9/llvm/test/Analysis/DXILMetadataAnalysis/entry-properties.ll) in my previous commit. Let me know if that suffices.

python3kgae replied:
We need to test the metadata generated in llvm ir is correct for this PR.
The test you mentioned only checks the dump of DXILMetadataAnalysis.

bharadwajy replied:
> We need to test the metadata generated in llvm ir is correct for this PR. The test you mentioned only checks the dump of DXILMetadataAnalysis.

I see. Pushed change with a new test. Thanks!

Commit:
Add a test to verify correct metadata creation for multiple
shader entries during library profile compilation.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+  return (uint32_t)Env - (uint32_t)llvm::Triple::Pixel;
+}
+
+struct ShaderEntryMDInfo : EntryProperties {
```
I think subclassing `EntryProperties` here makes this more complicated than necessary, and I'm not sure having a class for `ShaderEntryMDInfo` is really gaining much over a few static functions anyway. We only really have two methods in this struct that use anything from `EntryProperties`:
- constructEntryMetadata only needs a `Function *` and the metadata that's passed in.
- constructEntryPropMetadata could easily just take an `EntryProperties &` as a parameter.

bharadwajy replied:
> I think subclassing `EntryProperties` here makes this more complicated than necessary, and I'm not sure having a class for `ShaderEntryMDInfo` is really gaining much over a few static functions anyway. We only really have two methods in this struct that use anything from `EntryProperties`:
> 
> * constructEntryMetadata only needs a `Function *` and the metadata that's passed in.
> * constructEntryPropMetadata could easily just take an `EntryProperties &` as a parameter.

Deleted `struct ShaderEntryInfo`; refactored functionality in member functions as static functions.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+  case Triple::EnvironmentType::Library: {
+    // Library has an entry metadata with resource table metadata and all other
+    // MDNodes as null.
+    EntryProperties EP{};
+    // FIXME: ShaderFlagsAnalysis pass needs to collect and provide ShaderFlags
+    // for each entry function. Currently, ShaderFlags value provided by
+    // ShaderFlagsAnalysis pass is created by walking *all* the function
+    // instructions of the module. Is it is correct to use this value for
+    // metadata of the empty library entry?
+    ShaderEntryMDInfo EmptyFunEntryProps(EP, Ctx, MDAnalysisInfo.ShaderProfile,
+                                         MDResources, ShaderFlags);
+    MDTuple *EmptyMDT = EmptyFunEntryProps.getAsMetadata();
```
Given that the empty props are basically just references to the resources and module flags I think it'd be simpler to just emit them directly rather than do all of the work for `EntryProperties`/`ShaderEntryMDInfo` to handle nullptr for the `Entry` function.

bharadwajy replied:
> Given that the empty props are basically just references to the resources and module flags I think it'd be simpler to just emit them directly rather than do all of the work for `EntryProperties`/`ShaderEntryMDInfo` to handle nullptr for the `Entry` function.

Done.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    MDTuple *EmptyMDT = EmptyFunEntryProps.getAsMetadata();
+    EntryFnMDNodes.emplace_back(EmptyMDT);
+
+    for (auto EntryProp : MDAnalysisInfo.EntryPropertyVec) {
+      // FIXME: ShaderFlagsAnalysis pass needs to collect and provide
+      // ShaderFlags for each entry function. For now, assume shader flags value
+      // of entry functions being compiled for lib_* shader profile viz.,
+      // EntryPro.Entry is 0.
+      ShaderEntryMDInfo SEP(EntryProp, Ctx, MDAnalysisInfo.ShaderProfile,
+                            nullptr, 0);
+      MDTuple *MDT = SEP.getAsMetadata();
+      EntryFnMDNodes.emplace_back(MDT);
+    }
```
Wouldn't this loop be sufficient for all shader profiles? The logic for creating shader entries inside this loop is duplicated two times below with no apparent changes.

bharadwajy replied:
> Wouldn't this loop be sufficient for all shader profiles? The logic for creating shader entries inside this loop is duplicated two times below with no apparent changes.

Changed.

bogner commented on llvm/lib/Target/DirectX/DXILMetadata.cpp:
```diff
@@ -1,335 +0,0 @@
-    EntryList.emplace_back(&F);
-  }
-
-  // If there are no entries, do nothing. This is mostly to allow for writing
-  // tests with no actual entry functions.
-  if (EntryList.empty())
-    return;
```
It looks like you dropped this convenient behaviour and went back to crashing when we have no entry functions in a non-library shader.

bharadwajy replied:
> It looks like you dropped this convenient behaviour and went back to crashing when we have no entry functions in a non-library shader.

Fixed.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+  } break;
+  case Triple::EnvironmentType::Compute: {
+    size_t NumEntries = MDAnalysisInfo.EntryPropertyVec.size();
+    assert(NumEntries == 1 &&
+           "Compute shader: One and only one entry expected");
```
We should probably diagnose an error if we have the wrong number of entry points, rather than crashing.

bharadwajy replied:
> We should probably diagnose an error if we have the wrong number of entry points, rather than crashing.

Added `class DiagnosticinfoModuleFormat` and used it to report the error. Added a new test to verify that error is triggered.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+static void createEntryMD(Module &M, const uint64_t ShaderFlags,
+                          const dxil::ModuleMetadataInfo &MDAnalysisInfo) {
+  auto &Ctx = M.getContext();
+  // FIXME: generate metadata for resource.
+  MDTuple *MDResources = nullptr;
+  if (auto *NamedResources = M.getNamedMetadata("dx.resources"))
+    MDResources = dyn_cast<MDTuple>(NamedResources->getOperand(0));
```
This FIXME seems wrong - we're generating the resource metadata in emitResourceMetadata. Why not make that function return the `NamedMDNode` rather than looking it up again?

bharadwajy replied:
> This FIXME seems wrong - we're generating the resource metadata in emitResourceMetadata. Why not make that function return the `NamedMDNode` rather than looking it up again?

Changed per suggestion.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+                              const dxil::ModuleMetadataInfo &MDAnalysisInfo) {
+  LLVMContext &Ctx = M.getContext();
+  IRBuilder<> IRB(Ctx);
+  if (MDAnalysisInfo.ValidatorVersion.empty()) {
```
This pass shouldn't know the details of how the analysis gets the validator version, and so it should really emit/update the `dx.valver` version to whatever the metadata analysis says. Also IIRC we decided to set the default in the frontend, so I'm not sure we should be duplicating that logic here.

bharadwajy replied:
> This pass shouldn't know the details of how the analysis gets the validator version, and so it should really emit/update the `dx.valver` version to whatever the metadata analysis says. Also IIRC we decided to set the default in the frontend, so I'm not sure we should be duplicating that logic here.

Changed.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
-  dxil::createShaderModelMD(M);
-  dxil::createDXILVersionMD(M);
+                              const ComputedShaderFlags &ShaderFlags,
+                              const dxil::ModuleMetadataInfo &MDAnalysisInfo) {
```
Some of the names here are unusual because of this pass's legacy, but we typically abbreviate the analysis results to match the type name (see `DRM` above). This could be `const dxil::ModuleMetadataInfo &MMI`.

bharadwajy replied:
> Some of the names here are unusual because of this pass's legacy, but we typically abbreviate the analysis results to match the type name (see `DRM` above). This could be `const dxil::ModuleMetadataInfo &MMI`.

Changed.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+                              const dxil::ModuleMetadataInfo &MDAnalysisInfo) {
+  LLVMContext &Ctx = M.getContext();
+  IRBuilder<> IRB(Ctx);
+  if (MDAnalysisInfo.ValidatorVersion.empty()) {
+    // Module has no metadata node signifying valid validator version.
+    // Create metadata dx.valver node with version value of 1.0
+    const VersionTuple DefaultValidatorVer{1, 0};
+    Metadata *MDVals[2];
+    MDVals[0] =
+        ConstantAsMetadata::get(IRB.getInt32(DefaultValidatorVer.getMajor()));
+    MDVals[1] = ConstantAsMetadata::get(
+        IRB.getInt32(DefaultValidatorVer.getMinor().value_or(0)));
+    NamedMDNode *ValVerNode = M.getOrInsertNamedMetadata("dx.valver");
+    ValVerNode->addOperand(MDNode::get(Ctx, MDVals));
+  }
+
+  Metadata *SMVals[3];
+  VersionTuple SM = MDAnalysisInfo.ShaderModelVersion;
+  SMVals[0] =
+      MDString::get(Ctx, getShortShaderStage(MDAnalysisInfo.ShaderProfile));
+  SMVals[1] = ConstantAsMetadata::get(IRB.getInt32(SM.getMajor()));
+  SMVals[2] = ConstantAsMetadata::get(IRB.getInt32(SM.getMinor().value_or(0)));
+  NamedMDNode *SMMDNode = M.getOrInsertNamedMetadata("dx.shaderModel");
+  SMMDNode->addOperand(MDNode::get(Ctx, SMVals));
+
+  VersionTuple DXILVer = MDAnalysisInfo.DXILVersion;
+  Metadata *DXILVals[2];
+  DXILVals[0] = ConstantAsMetadata::get(IRB.getInt32(DXILVer.getMajor()));
+  DXILVals[1] =
+      ConstantAsMetadata::get(IRB.getInt32(DXILVer.getMinor().value_or(0)));
+  NamedMDNode *DXILVerMDNode = M.getOrInsertNamedMetadata("dx.version");
+  DXILVerMDNode->addOperand(MDNode::get(Ctx, DXILVals));
```
I'd probably move these three blocks of logic into simple functions to keep the level of abstraction at a single level here, so this function would just look something like:
```c++
  emitValidatorVersionMD(M, MMI);
  emitShaderModuleMD(M, MMI);
  emitDXILVersionMD(M, MMI);
  NamedMDNode *ResourceNode = emitResourceMetadata(M, DRM, MDResources);
  
  SmallVector<MDNode *> EntryNodes;
  if (MMI.ShaderProfile == Triple::EnvironmentType::Library)
    EntryNodes.push_back(emitTopLevelLibraryNode(M, ResourceNode, ShaderFlags));
  else if (MMI.EntryProps.size() != 1)
    diagnoseWrongNumberOfEntryProps(MMI.ShaderProfile, MMI.EntryProps.size());
  for (const EntryProperties &Entry : MMI.EntryProps)
    EntryNodes.push_back(emitEntryNode(M, Entry, MMI, ResourceNode, ShaderFlags));
  emitEntryPointsMD(M, EntryNodes);
```

bharadwajy replied:
> I'd probably move these three blocks of logic into simple functions to keep the level of abstraction at a single level here, so this function would just look something like:
> 
> ```c++
>   emitValidatorVersionMD(M, MMI);
>   emitShaderModuleMD(M, MMI);
>   emitDXILVersionMD(M, MMI);
>   NamedMDNode *ResourceNode = emitResourceMetadata(M, DRM, MDResources);
>   
>   SmallVector<MDNode *> EntryNodes;
>   if (MMI.ShaderProfile == Triple::EnvironmentType::Library)
>     EntryNodes.push_back(emitTopLevelLibraryNode(M, ResourceNode, ShaderFlags));
>   else if (MMI.EntryProps.size() != 1)
>     diagnoseWrongNumberOfEntryProps(MMI.ShaderProfile, MMI.EntryProps.size());
>   for (const EntryProperties &Entry : MMI.EntryProps)
>     EntryNodes.push_back(emitEntryNode(M, Entry, MMI, ResourceNode, ShaderFlags));
>   emitEntryPointsMD(M, EntryNodes);
> ```

Done.

bogner commented on llvm/lib/Target/DirectX/DirectXTargetMachine.cpp:
```diff
@@ -50,6 +51,7 @@ extern "C" LLVM_EXTERNAL_VISIBILITY void LLVMInitializeDirectXTarget() {
   initializeDXContainerGlobalsPass(*PR);
   initializeDXILOpLoweringLegacyPass(*PR);
   initializeDXILTranslateMetadataLegacyPass(*PR);
+  initializeDXILMetadataAnalysisWrapperPassPass(*PR);
```
We shouldn't need this - this pass isn't part of the DirectX target.

bharadwajy replied:
> We shouldn't need this - this pass isn't part of the DirectX target.

Deleted.

bogner commented on llvm/test/CodeGen/DirectX/CreateHandle.ll:
```diff
@@ -14,7 +14,7 @@ target triple = "dxil-pc-shadermodel6.0-compute"
 declare i32 @some_val();
 
-define void @test_buffers() {
+define void @test_buffers() #0 {
```
If we allow zero entry points in tests we don't need this.

bharadwajy replied:
> If we allow zero entry points in tests we don't need this.

Reverted.

Commit:
Address PR feedback

- Delete derived struct ShaderEntryMDInfo and move the functionlaity
  in its methods as static functions.
- Add class DiagnosticInfoModuleFormat derived from DiagnosticInfo for
  diagnostics reporting in TranslateMetadata pass
- Consume Resource metadata information constructed by emitResourceMetadata()
- Move generation of metadata for validator version, Shader Model
  version, DXIL Version into separate functions.
- Changes to accept input shader modules with no entry functions
- Update/emit named metadata dx.valver only if Metadata Analysis info
  contains the information; it is not created to with a default value,
  if it is not.

bharadwajy commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    break;
+  }
+  llvm_unreachable("Unsupported environment for DXIL generation.");
+  return "";
```
> No need for `return` after `llvm_unreachable`.

Deleted.

python3kgae commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +94,251 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    uint64_t EntryShaderFlags =
+        (MMDI.ShaderProfile == Triple::EnvironmentType::Library) ? 0
+                                                                 : ShaderFlags;
+    EntryFnMDNodes.emplace_back(emitEntryMD(EntryProp, Signatures, ResourceMD,
```
Need to make sure EP.ShaderStage == MMDI.ShaderProfile when MMDI.ShaderProfile != Triple::EnvironmentType::Library

bharadwajy replied:
> Need to make sure EP.ShaderStage == MMDI.ShaderProfile when MMDI.ShaderProfile != Triple::EnvironmentType::Library

Pushed changes along with a test.

Commit:
Add a check to ensure entry shader stage is the same as target profile
for non-library target profile compilation.

Add a test to verify the check triggers diagnostic report appropriately.

Commit:
Changes upon rebasing

damyanp commented on llvm/lib/Target/DirectX/DXILPrepare.cpp:
```diff
@@ -247,9 +247,10 @@ class DXILPrepareModule : public ModulePass {
     AU.addPreserved<ShaderFlagsAnalysisWrapper>();
     AU.addPreserved<DXILResourceMDWrapper>();
-    AU.addPreserved<DXILMetadataAnalysisWrapperPass>();
+    AU.addRequired<DXILMetadataAnalysisWrapperPass>();
```
Likely an opportunity for me to learn something:

If we're using `setPreservesAll()` why do we need to explicitly call `addPreserved` on any other passes?

bogner replied:
Is it correct to say that this preserves all? I'm not sure if any analyses cache information about function attributes, so maybe it's okay there, but we do modify the IR so I doubt it's safe to really say we preserve all analyses...

bharadwajy replied:
> Is it correct to say that this preserves all? I'm not sure if any analyses cache information about function attributes, so maybe it's okay there, but we do modify the IR so I doubt it's safe to really say we preserve all analyses...

I share the concern about specifying `setPreservedAll()` for `DXILPrepare` pass that actually changes the LLVM IR (as I was discussing offline with @damyanp earlier). As things stand, `DXContainerGlobals` pass that runs after `DXILPrepare` pass uses `DXILMetadataAnalysis` info (i.e., requires the pass) but running it before `DXContainerGlobals` would produce incomplete info since `DXILPrepare` deletes function attributes. Hence the unsatisfactory workaround of `setPreservedAll()` in `DXILPrepare`. 

To address this, I'd like to move attribute removal out of `DXILPrepare` into `DXContainerGlobals` pass and (correctly) delete `setPreservedAll()` from `DXILPrepare` pass. Is there a known reason for removing function attributes in `DXILPrepare` pass and not in a later pass?

bharadwajy replied:
I'll plan to move the attribute removal in a follow on PR, if there are no technical reasons to address the issue raised here.

bogner replied:
I still don't understand why we would change this to `setPreservesAll`. If you just need to preserve `DXILMetadataAnalysis` then the existing `AU.addPreserved<DXILMetadataAnalysisWrapperPass>()` call should do that, no?

bogner replied:
I'm not arguing that the `addRequired<DXILMetadataAnalysisWrapperPass>()` isn't necessary, it obviously is. I'm saying we shouldn't change the specific list of preserved passes to `setPreservesAll`. This pass does not preserve all analyses, since it modifies instructions. The code before your change looks correct to me:
```c++
    AU.addPreserved<ShaderFlagsAnalysisWrapper>();
    AU.addPreserved<DXILResourceMDWrapper>();
    AU.addPreserved<DXILMetadataAnalysisWrapperPass>();
    AU.addPreserved<DXILResourceWrapperPass>();
```

We can simply add `AU.addRequired<DXILMetadataAnalysisWrapperPass>();` after those calls instead of changing those calls to `setPreservesAll`, since that is incorrect.

bharadwajy replied:
Changed.

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -7,26 +7,62 @@
+/// A simple Wrapper DiagnosticInfo that generates Module-level diagnostic
+class DiagnosticInfoModuleFormat : public DiagnosticInfo {
+private:
+  Twine Msg;
```
```suggestion
  const Twine &Msg;
```

damyanp replied:
Although I'm wondering why this can't just be a `StringRef`?

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -7,26 +7,62 @@
+                             DiagnosticSeverity Severity = DS_Error)
+      : DiagnosticInfo(DK_Unsupported, Severity), Msg(Msg), Mod(M) {}
+
+  static bool classof(const DiagnosticInfo *DI) {
+    return DI->getKind() == DK_Unsupported;
+  }
+
+  const Twine &getMessage() const { return Msg; }
```
If I delete `classof` and `getMessage` this code still compiles. Are they really needed?

bharadwajy replied:
> If I delete `classof` and `getMessage` this code still compiles. Are they really needed?

They are not currently used. The class was created following the way various derived classes of `DiagnosticInfo` in `DiagnosticInfo.h` - expecting that these methods may be of use if/when this diagnostic is used more. Deleted for now.

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +94,258 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+        ConstantInt::get(Type::getInt32Ty(Ctx), Value)));
+    break;
+  default:
+    assert(false && "NYI: Unhandled entry property tag");
```
```suggestion
    llvm_unreachable("NYI: Unhandled entry property tag");
```

bharadwajy replied:
> 

Changed.

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +94,258 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    if (EP.ShaderStage == Triple::EnvironmentType::Compute) {
+      MDVals.emplace_back(ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt32Ty(Ctx), NumThreadsTag)));
+      std::vector<Metadata *> NumThreadVals;
```
```suggestion
      SmallVector<Metadata *, 3> NumThreadVals;
```

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +94,258 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    if (EP.ShaderStage == Triple::EnvironmentType::Compute) {
+      MDVals.emplace_back(ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt32Ty(Ctx), NumThreadsTag)));
+      std::vector<Metadata *> NumThreadVals;
+      NumThreadVals.emplace_back(ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt32Ty(Ctx), EP.NumThreadsX)));
+      NumThreadVals.emplace_back(ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt32Ty(Ctx), EP.NumThreadsY)));
+      NumThreadVals.emplace_back(ConstantAsMetadata::get(
+          ConstantInt::get(Type::getInt32Ty(Ctx), EP.NumThreadsZ)));
+      MDVals.emplace_back(MDNode::get(Ctx, NumThreadVals));
```
```suggestion
      Metadata *NumThreadVals[] = {
          ConstantAsMetadata::get(
              ConstantInt::get(Type::getInt32Ty(Ctx), EP.NumThreadsX)),
          ConstantAsMetadata::get(
              ConstantInt::get(Type::getInt32Ty(Ctx), EP.NumThreadsY)),
          ConstantAsMetadata::get(
              ConstantInt::get(Type::getInt32Ty(Ctx), EP.NumThreadsZ))};
      MDVals.emplace_back(MDNode::get(Ctx, NumThreadVals));
```

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +94,258 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+        emitTopLevelLibraryNode(M, ResourceMD, ShaderFlags));
+  else if (MMDI.EntryPropertyVec.size() > 1) {
+    M.getContext().diagnose(DiagnosticInfoModuleFormat(
+        M, "Non-library shader: One and only one entry expected"));
```
Do we have a better term than non-library shader we could use here?

bharadwajy replied:
> Do we have a better term than non-library shader we could use here?

I am open to a different term/wording that conveys one and only one entry is expected in shaders being compiled for a target profile other than `library`.

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +94,258 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    if (MMDI.ShaderProfile != Triple::EnvironmentType::Library) {
+      if (EntryProp.ShaderStage != MMDI.ShaderProfile) {
+        M.getContext().diagnose(DiagnosticInfoModuleFormat(
+            M, "Non-library shader: Stage of Shader entry different from "
```
"Stage of Shader entry" - the code and diagnostics seem to be inconsistent about when "shader profile" and "shader stage" are used.

bharadwajy replied:
> "Stage of Shader entry" - the code and diagnostics seem to be inconsistent about when "shader profile" and "shader stage" are used.

"Shader Profile" is expected to convey the target shader profile while "Shader Stage" is associated with the entry. Modified the message to "Non-library shader: Stage of shader entry different from shader target profile"

damyanp replied:
> "Non-library shader: Stage of shader entry different from shader target profile"

So, I think: if this diagnostic is emitted for IR generated by Clang then this is a bug in Clang.  Does that sound right?  The only way to really hit this is for some other frontend to do the wrong thing?

Given that the case we can probably frame the diagnostic using llvm terms like "environment" and "triple" if that helps at all?

bharadwajy replied:
> > "Non-library shader: Stage of shader entry different from shader target profile"
> 
> So, I think: if this diagnostic is emitted for IR generated by Clang then this is a bug in Clang. Does that sound right? The only way to really hit this is for some other frontend to do the wrong thing?
> 
> Given that the case we can probably frame the diagnostic using llvm terms like "environment" and "triple" if that helps at all?

How about - 
```
Shader stage '<entry-shader-stage>' of entry '<entry>' different from specified target profile '<target-shader-profile>'
```
 where the exact mismatching component is spelt out? An instance of such a message would look like  

```
Shader stage 'cs' of entry 'cs_entry' different from specified target profile 'pixel'
```

damyanp reviewed (commented):
LGTM, but I'll leave approval to those with more on the ground knowledge of how this works.  Some comments around some code style / cleanups.

Commit:
Address PR feedback

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -34,7 +35,7 @@ using namespace llvm::dxil;
 class DiagnosticInfoModuleFormat : public DiagnosticInfo {
 private:
-  Twine Msg;
+  const Twine Msg;
```
```suggestion
  const Twine &Msg;
```

bharadwajy replied:
Changed.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -7,26 +7,57 @@
-static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
-                                 const dxil::Resources &MDResources) {
+/// A simple Wrapper DiagnosticInfo that generates Module-level diagnostic
+class DiagnosticInfoModuleFormat : public DiagnosticInfo {
```
This needs to be wrapped in an anonymous namespace to avoid leaking the name / risking name collisions. Also I don't really understand the name - what does the word format here mean?

bharadwajy replied:
> This needs to be wrapped in an anonymous namespace to avoid leaking the name / risking name collisions. Also I don't really understand the name - what does the word format here mean?

Done.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -7,26 +7,57 @@
+      : DiagnosticInfo(DK_Unsupported, Severity), Msg(Msg), Mod(M) {}
+
+  void print(DiagnosticPrinter &DP) const override {
+    std::string Str;
+    raw_string_ostream OS(Str);
+
+    OS << Mod.getName() << ": " << Msg << '\n';
+    OS.flush();
+    DP << Str;
```
Why use a string here rather than printing into the `DiagnosticPrinter` directly?

bharadwajy replied:
> Why use a string here rather than printing into the `DiagnosticPrinter` directly?

Changed.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +89,257 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+}
+
+namespace {
+enum EntryPropsTag {
+  ShaderFlagsTag = 0,
+  GSStateTag,
+  DSStateTag,
+  HSStateTag,
+  NumThreadsTag,
+  AutoBindingSpaceTag,
+  RayPayloadSizeTag,
+  RayAttribSizeTag,
+  ShaderKindTag,
+  MSStateTag,
+  ASStateTag,
+  WaveSizeTag,
+  EntryRootSigTag,
+};
```
I think it'd be better to use `enum class` here. This would also mean the "Tag" suffix on each value isn't necessary since it doesn't pollute the local namespace that way.

bharadwajy replied:
> I think it'd be better to use `enum class` here. This would also mean the "Tag" suffix on each value isn't necessary since it doesn't pollute the local namespace that way.

Done.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +89,257 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+    MDVals.emplace_back(ConstantAsMetadata::get(
+        ConstantInt::get(Type::getInt32Ty(Ctx), Value)));
+    break;
+  default:
+    llvm_unreachable("NYI: Unhandled entry property tag");
```
It'd be better to spell out all of the cases so that we get a covered switch diagnostic if one is missing. I'd even argue it might be better to leave the unimplemented ones out of the enum until we support them so it's easier to see where all we need to make updates, though that is a little bit awkward since we'd need gaps in the enum for now.

bharadwajy replied:
> It'd be better to spell out all of the cases so that we get a covered switch diagnostic if one is missing. I'd even argue it might be better to leave the unimplemented ones out of the enum until we support them so it's easier to see where all we need to make updates, though that is a little bit awkward since we'd need gaps in the enum for now.

Done.

bharadwajy replied:
> It'd be better to spell out all of the cases so that we get a covered switch diagnostic if one is missing. I'd even argue it might be better to leave the unimplemented ones out of the enum until we support them so it's easier to see where all we need to make updates, though that is a little bit awkward since we'd need gaps in the enum for now.

I did leave the unimplemented enums in with an intent to keep them on record to add future support. I am not opposed to deleting them if there is a strong objection or some LLVM convention for such situations.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +89,257 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+  return MDNode::get(Ctx, MDVals);
+}
+
+// Each entry point metadata record specifies:
+//  * reference to the entry point function global symbol
+//  * unmangled name
+//  * list of signatures
+//  * list of resources
+//  * list of tag-value pairs of shader capabilities and other properties
```
It'd probably be better to put this comment inside the function below where we're creating a metadata node with these values in it.

bharadwajy replied:
> It'd probably be better to put this comment inside the function below where we're creating a metadata node with these values in it.

Done.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -58,25 +89,257 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+}
+
+static void emitValidatorVersionMD(Module &M, const ModuleMetadataInfo &MMDI) {
+  if (!MMDI.ValidatorVersion.empty()) {
```
It doesn't make a big difference here, but it's best to use early returns to avoid unnecessary nesting. So,
```c++
  if (MMDI.ValidatorVersion.empty())
    return;
```

Commit:
Address more PR feedback
- Rename class DiagnosticInfoModuleFormat as class DiagnosticInfoTranslateMD
  and enclose it in an anonymous namespace.
- Print message directly to DiagnosticPrinter in DiagnosticInfoTranslateMD::print()
- Change enum EntryPropsTag to scop[ed enum class
- List all ENtryPropsTags in case statement
- Moved comment inside constructEntryMetadata()
- Changed emitValidatorVersionMD() to return early as appropriate.

Commit:
Early return and movement of comment changes missed in earlier commit

Commit:
Restore preserved pass specification in DXILPrepare
Add precise paases preserved in DXILTranslateMetadata instead of setPreservesAll()
